{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "coveragefile = 'variants/coverage.tsv'\n",
    "input_qafile = 'qa.csv'\n",
    "output_qafile = 'qa_vals.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.core.display import display, HTML\n",
    "display(HTML(\n",
    "    '<style>'\n",
    "        '#notebook { padding-top:0px !important; } ' \n",
    "        '.container { width:100% !important; } '\n",
    "        '.end_space { min-height:0px !important; } '\n",
    "    '</style>'\n",
    "))\n",
    "\n",
    "display(HTML('''<script>\n",
    "$(document).ready(function() {\n",
    "  \n",
    "  all_cells = $('div.cell.border-box-sizing.code_cell.rendered')\n",
    "  \n",
    "  $.each(all_cells, function(idx, val){\n",
    "      if ($(val).find('div.output_wrapper').length === 0){\n",
    "        $(val).hide();\n",
    "      }\n",
    "    });\n",
    "  });\n",
    "  \n",
    "</script>\n",
    "'''))\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.colors import ListedColormap\n",
    "import seaborn as sns\n",
    "sns.set_style('ticks')\n",
    "\n",
    "import plotly.graph_objs as go\n",
    "from plotly.offline import init_notebook_mode, iplot\n",
    "import plotly.figure_factory as ff\n",
    "import plotly.io as pio\n",
    "\n",
    "\n",
    "init_notebook_mode()\n",
    "\n",
    "from ipywidgets import interact\n",
    "from IPython.display import display, Markdown, Latex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "def extract_sample_conditions(samples):\n",
    "    '''Function to parse a list of sample names and return their condition.\n",
    "    Parameters:\n",
    "        samples: list of sample names\n",
    "    Returns:\n",
    "        (condition, condition_extended): tuple of np.arrays. \n",
    "        condition is a np.array where 0 = negative control,\n",
    "        1 = positive control, 2 = experimental condition, -1 = unassigned.  \n",
    "        condition_extended is a np.array where \n",
    "        0 = H2O negative control, 1 = empty well negative control, 2 = negative PCR tests,\n",
    "        3 = negative controls of Functional Genomics ZH, \n",
    "        4 = Twist positive control, 5 = Positive Controls (TO CONFIRM),\n",
    "        6 = ETHZ_ID sample, 7 = wastewater sample, 8 = functional genomics center sample, \n",
    "        9= Basel Sequencing of UZH Virology lab, 10 = Labormedizinisches Zentrum Dr Risch, 11 = USZ Tier sample,\n",
    "        -1 = unassigned\n",
    "    '''\n",
    "    # make a list of regexps\n",
    "    re_list = [\n",
    "    # regexp for H2O negative controls eg: H2O_CP002_A7\n",
    "    re.compile(\"^H2O\"),\n",
    "    # regexp for Empty wells negative controls eg: EMPTY_CP002_A11\n",
    "    re.compile(\"^((EMPTY)|(empty))\"),\n",
    "    # regexp for samples of negative tests eg: neg_109_B2\n",
    "    re.compile(\"^neg_\"),\n",
    "    # regexp for negative controls of Functional Genomics ZH eg: NTC_NA_NTC_NA\n",
    "    re.compile(\"^NTC_NA_NTC_NA\"),\n",
    "    # regexp for samples of Twist positive controls eg: pos_MN908947_3_1_100000_CP0002\n",
    "    re.compile(\"^(pos_)|(Twist_control)\"),\n",
    "    # regexp for samples of Positive Controls (TO CONFIRM) eg: CoV_ctrl_1_1_10000\n",
    "    re.compile(\"CoV_ctrl_\"),\n",
    "    # regexp for ETH ID samples eg: 160000_434_D02\n",
    "    re.compile(\"^[0-9]{6}(_Plate){0,1}_(p){0,1}[0-9]+_\"),\n",
    "    # regexp for wastewater samples eg: 09_2020_03_24_B\n",
    "    re.compile(\"^[0-9]{2}_202[0-9]_\"),\n",
    "    # regexp for functional genomics eg: 30430668_Plate_8_041120tb3_D7\n",
    "    re.compile(\"^[0-9]{8}_Plate_[0-9]+\"),\n",
    "    # regexp for Basel sequencing for some UZH lab (UZH Virology ?) eg: A2_722\n",
    "    re.compile(\"^[A-Z][0-9]_[0-9]+\"),\n",
    "    # regexp for lone sample by the lab \"Labormedizinisches Zentrum Dr Risch\": 674597001\n",
    "    re.compile(\"^674597001\"),\n",
    "    # regexp for USZ Tier eg: USZ_5_Tier\n",
    "    re.compile(\"^USZ_[0-9]_Tier\")\n",
    "    ]\n",
    "\n",
    "    # make 0-1 array with n_samples lines and p_conditions columns\n",
    "    re_matches = [[int(regex.match(sam) is not None) for sam in samples] for regex in re_list]\n",
    "    re_matches_array = np.array(re_matches)\n",
    "\n",
    "    # check that samples are assigned to 1 condition max\n",
    "    if np.max(np.sum(re_matches_array, axis=0)) > 1:\n",
    "        print(\"Warning: some samples were assigned to multiple conditions.\")\n",
    "\n",
    "    # assign to each its extended condition\n",
    "    condition_extended = np.argmax(re_matches_array, axis=0)\n",
    "    condition_extended = condition_extended - (np.sum(re_matches_array, axis=0) == 0) # set unassigned to -1\n",
    "\n",
    "    # assign to each its condition\n",
    "    condition = np.in1d(condition_extended, [0,1,2,3])*0\n",
    "    condition = condition + np.in1d(condition_extended, [4,5])*1\n",
    "    condition = condition + np.in1d(condition_extended, [6,7,8,9,10,11])*2\n",
    "    condition = condition + np.in1d(condition_extended, [-1])*(-1)\n",
    "    \n",
    "    return (condition, condition_extended)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.read_csv(input_qafile)\n",
    "data = data.set_index('sample')\n",
    "data = data.reset_index().rename(columns={'index':'sample'})\n",
    "\n",
    "coverage = pd.read_csv(coveragefile, sep='\\t',index_col=['ref','pos']).droplevel('ref',axis=0)\n",
    "coverage = coverage.T\n",
    "\n",
    "data['input_reads'] = data['input_r1'] + data['input_r2']\n",
    "data['goodreads'] = data['goodpairs'] * 2\n",
    "data['conditions'], data['conditions_extended'] = extract_sample_conditions(data['sample'])",
    "\n",
    "data['group'] = 'Real'\n",
    "data.loc[data.conditions_extended == 0] = 'H2O'\n",
    "data.loc[data.conditions_extended == 1] = 'Empty'\n",
    "data.loc[data.conditions_extended == 2] = 'Negative'\n",
    "data.loc[data.conditions == 2] = 'Positive'\n",
    "\n",
    "real_samples = data.loc[data.conditions == 2]\n",
    "h2o_samples = data.loc[data.conditions_extended == 0]\n",
    "emp_samples = data.loc[data.conditions_extended == 1]\n",
    "neg_samples = data.loc[data.conditions_extended == 2]\n",
    "pos_samples = data.loc[data.conditions == 2]\n",
    "\n",
    "subsets_dict = dict(All=data, Real=real_samples, Empty=emp_samples, H2O=h2o_samples, Negative=neg_samples, \n",
    "                    Positive=pos_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sizes = dict()\n",
    "for key in subsets_dict:\n",
    "    sizes[key] = subsets_dict[key].shape[0]\n",
    "sizes = pd.DataFrame(sizes, index=range(1))\n",
    "\n",
    "groupsize = sizes.T.reset_index().rename(columns={'index': 'group', 0: 'size'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "coverage_quartiles = coverage.T.describe().loc[['25%', '50%', '75%']]\n",
    "\n",
    "coverage_quartiles = coverage_quartiles.T.reset_index().rename(columns={'index': 'sample'})\n",
    "\n",
    "coverage_quartiles['sample'] = coverage_quartiles['sample'].apply(lambda s: s.split('-')[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "coverage = coverage.reset_index(drop=True)\n",
    "real_mean = coverage.loc[data['group']=='Real'].quantile(q=0.5, axis=0)\n",
    "real_1q = coverage.loc[data['group']=='Real'].quantile(q=0.25, axis=0)\n",
    "real_3q = coverage.loc[data['group']=='Real'].quantile(q=0.75, axis=0)\n",
    "empty_mean = coverage.loc[data['group']=='Empty'].quantile(q=0.5, axis=0)\n",
    "empty_1q = coverage.loc[data['group']=='Empty'].quantile(q=0.25, axis=0)\n",
    "empty_3q = coverage.loc[data['group']=='Empty'].quantile(q=0.75, axis=0)\n",
    "h2o_mean = coverage.loc[data['group']=='H2O'].quantile(q=0.5, axis=0)\n",
    "h2o_1q = coverage.loc[data['group']=='H2O'].quantile(q=0.25, axis=0)\n",
    "h2o_3q = coverage.loc[data['group']=='H2O'].quantile(q=0.75, axis=0)\n",
    "neg_mean = coverage.loc[data['group']=='Negative'].quantile(q=0.5, axis=0)\n",
    "pos_mean = coverage.loc[data['group']=='Positive'].quantile(q=0.5, axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "coverage_pos_percentiles = pd.DataFrame(columns={'real'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "quantity = 'basequal'\n",
    "read_no = 1\n",
    "quantity = f'r{read_no}_{quantity}'\n",
    "\n",
    "percs = dict()\n",
    "for key in subsets_dict:\n",
    "    percs[key] = subsets_dict[key][f'fastqc_{quantity}'].value_counts(normalize=True)\n",
    "\n",
    "fqcpercs = pd.DataFrame(percs).fillna(0) * 100\n",
    "#fqcpercs = fqcpercs.loc[['PASS','WARNING','FAIL','']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "upper_bound_reals = 25\n",
    "lower_bound_controls = 90 \n",
    "\n",
    "percs = dict()\n",
    "nums = dict()\n",
    "for key in subsets_dict:\n",
    "    percs[key] = (1. - subsets_dict[key]['goodpairs']/(subsets_dict[key][['input_r1', 'input_r2']].sum(axis=1)/2)) * 100\n",
    "    nums[key] = (subsets_dict[key][['input_r1', 'input_r2']].sum()/2) - subsets_dict[key]['goodpairs']\n",
    "    \n",
    "percs = pd.DataFrame(percs)\n",
    "nums = pd.DataFrame(nums)\n",
    "\n",
    "rejreads = percs\n",
    "rejreads['sample'] = data['sample']\n",
    "rejreads['Absolute'] = ((data[['input_r1', 'input_r2']].sum(axis=1)/2) - data['goodpairs']).astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lower_bound_reals = 90\n",
    "upper_bound_controls = 10\n",
    "\n",
    "percs = dict()\n",
    "for key in subsets_dict:\n",
    "    percs[key] = subsets_dict[key]['alnreads'] / (subsets_dict[key]['goodpairs']*2.) * 100\n",
    "\n",
    "percs = pd.DataFrame(percs)\n",
    "\n",
    "alnreads = percs\n",
    "alnreads['sample'] = data['sample']\n",
    "alnreads['Absolute'] = data['alnreads']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bwa_insert_dist = data[['sample', 'bwa_insert_min', 'bwa_insert_mean', 'bwa_insert_max']]\n",
    "bwa_insert_dist = bwa_insert_dist.rename(columns={'bwa_insert_min':'min', 'bwa_insert_mean':'mean', 'bwa_insert_max':'max'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vals_N = dict()\n",
    "vals_lc = dict()\n",
    "for key in subsets_dict:\n",
    "    vals_N[key] = subsets_dict[key]['consensus_N']\n",
    "    vals_lc[key] = subsets_dict[key]['consensus_lower']\n",
    "\n",
    "vals_N = pd.DataFrame(vals_N)\n",
    "vals_lc = pd.DataFrame(vals_lc)\n",
    "\n",
    "lcbases = vals_lc\n",
    "lcbases['sample'] = data['sample']\n",
    "\n",
    "nbases = vals_N\n",
    "nbases['sample'] = data['sample']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "whole_genome_length = 29903\n",
    "vals = dict()\n",
    "for key in subsets_dict:\n",
    "    vals[key] = whole_genome_length - subsets_dict[key]['consensus_N'] - subsets_dict[key]['match_id']\n",
    "    \n",
    "vals = pd.DataFrame(vals)\n",
    "\n",
    "diffbases = vals\n",
    "diffbases['sample'] = data['sample']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vals_snv = dict()\n",
    "vals_f = dict()\n",
    "vals_maj = dict()\n",
    "for key in subsets_dict:\n",
    "    vals_snv[key] = subsets_dict[key]['shorah_snv']\n",
    "    vals_f[key] = subsets_dict[key]['shorah_filtered']\n",
    "    vals_maj[key] = subsets_dict[key]['shorah_majority']\n",
    "\n",
    "vals_snv = pd.DataFrame(vals_snv)\n",
    "vals_f = pd.DataFrame(vals_f)\n",
    "vals_maj = pd.DataFrame(vals_maj)\n",
    "\n",
    "filtered_snvs = vals_f\n",
    "filtered_snvs['sample'] = data['sample']\n",
    "maj_snvs = vals_maj\n",
    "maj_snvs['sample'] = data['sample']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "flags = pd.DataFrame({'sample':data['sample']})\n",
    "flags_values = pd.DataFrame({'sample':data['sample']})\n",
    "\n",
    "flags['coverage'] = 'PASS'\n",
    "assert (coverage_quartiles['sample'] == flags['sample']).all()\n",
    "coverage_3q_min_warning = 2000\n",
    "coverage_3q_min_fail = 1000\n",
    "coverage_failed = coverage_quartiles.loc[coverage_quartiles['75%'] < coverage_3q_min_warning]['sample'].index\n",
    "flags.loc[coverage_failed, 'coverage'] = 'WARNING'\n",
    "coverage_failed = coverage_quartiles.loc[coverage_quartiles['75%'] < coverage_3q_min_fail]['sample'].index\n",
    "flags.loc[coverage_failed, 'coverage'] = 'FAIL'\n",
    "flags_values['coverage'] = coverage_quartiles['75%']\n",
    "\n",
    "flags['r1_basequal'] = data['fastqc_r1_basequal']\n",
    "flags['r2_basequal'] = data['fastqc_r2_basequal']\n",
    "flags_values['r1_basequal'] = flags['r1_basequal']\n",
    "flags_values['r2_basequal'] = flags['r2_basequal']\n",
    "\n",
    "flags['rejreads'] = 'PASS'\n",
    "rejupper_bound_warning = 20\n",
    "rejupper_bound_fail = 40 # or minimum of negative controls?\n",
    "rejreads_failed = rejreads.loc[rejreads['All'] > rejupper_bound_warning].index\n",
    "flags.loc[rejreads_failed, 'rejreads'] = 'WARNING'\n",
    "rejreads_failed = rejreads.loc[rejreads['All'] > rejupper_bound_fail].index\n",
    "flags.loc[rejreads_failed, 'rejreads'] = 'FAIL'\n",
    "flags_values['rejreads'] = rejreads['All']\n",
    "\n",
    "flags['alnreads'] = 'PASS'\n",
    "alnlower_bound = 90\n",
    "alnreads_failed = alnreads.loc[alnreads['All'] < alnlower_bound].index\n",
    "flags.loc[alnreads_failed, 'alnreads'] = 'FAIL'\n",
    "flags_values['alnreads'] = alnreads['All']\n",
    "\n",
    "flags['insertsize'] = 'PASS'\n",
    "insupper_bound = 500\n",
    "inslower_bound = 300\n",
    "insertsize_failed = bwa_insert_dist.loc[(bwa_insert_dist['min'] < inslower_bound) | (bwa_insert_dist['max'] > insupper_bound)].index\n",
    "flags.loc[insertsize_failed, 'insertsize'] = 'FAIL'\n",
    "flags_values['insertsize'] = bwa_insert_dist['mean']\n",
    "\n",
    "flags['consensus_N'] = 'PASS'\n",
    "undupper_bound = 10000\n",
    "nbases_failed = nbases.loc[nbases['All'] > undupper_bound].index\n",
    "flags.loc[nbases_failed, 'consensus_N'] = 'FAIL'\n",
    "flags_values['consensus_N'] = nbases['All']\n",
    "\n",
    "flags['consensus_lcbases'] = 'PASS'\n",
    "ambupper_bound = 20000\n",
    "lcbases_failed = lcbases.loc[lcbases['All'] > ambupper_bound].index\n",
    "flags.loc[lcbases_failed, 'consensus_lcbases'] = 'FAIL'\n",
    "flags_values['consensus_lcbases'] = lcbases['All']\n",
    "\n",
    "flags['consensus_diffbases'] = 'PASS'\n",
    "diffupper_bound_warning = 50\n",
    "diffupper_bound_fail = 100\n",
    "diffbases_warning = diffbases.loc[diffbases['All'] > diffupper_bound_warning].index\n",
    "flags.loc[diffbases_warning, 'consensus_diffbases'] = 'WARNING'\n",
    "diffbases_failed = diffbases.loc[diffbases['All'] > diffupper_bound_fail].index\n",
    "flags.loc[diffbases_failed, 'consensus_diffbases'] = 'FAIL'\n",
    "flags_values['consensus_diffbases'] = diffbases['All']\n",
    "\n",
    "flags = flags.rename(columns={'snvs_filtered': 'snvs'})\n",
    "flags['snvs'] = 'PASS'\n",
    "snvsupper_bound = 1000\n",
    "filtered_failed = filtered_snvs.loc[filtered_snvs['All'] > snvsupper_bound].index\n",
    "flags.loc[filtered_failed, 'snvs'] = 'FAIL'\n",
    "flags_values['snvs'] = filtered_snvs['All']\n",
    "\n",
    "flags['snvs_majority'] = 'PASS'\n",
    "majupper_bounds = np.max([np.ones(diffbases['All'].shape[0]), diffbases['All'].values], axis=0) * 10\n",
    "majlower_bounds = (np.max([np.ones(diffbases['All'].shape[0]), diffbases['All'].values], axis=0) * 0.1).astype(int)\n",
    "maj_failed = maj_snvs.loc[(maj_snvs['All'].values > majupper_bounds) | (maj_snvs['All'].values < majlower_bounds)].index\n",
    "flags.loc[maj_failed, 'snvs_majority'] = 'FAIL'\n",
    "flags_values['snvs_majority'] = maj_snvs['All']\n",
    "\n",
    "flags = flags.rename(columns={'sample': 'Sample'})\n",
    "flags_values = flags_values.rename(columns={'sample': 'Sample'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "splt=data['batch'].str.split(\"_\", n = 1, expand = True) \n",
    "data['date']=splt[0]\n",
    "data['flowcell']=splt[1]\n",
    "date = str(pd.to_datetime(data.loc[0, 'date'], format='%Y%m%d').date())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# List of real samples that failed for consensus usage\n",
    "# Sample is failed if any of the flags is a fail\n",
    "sample_status = pd.DataFrame(dict(samples=flags['Sample'], status='PASS'))\n",
    "for index, row in flags.iterrows():\n",
    "    if np.any(np.array(row.values[:-2]) == 'FAIL'):\n",
    "        sample_status.loc[index,'status'] = 'FAIL'\n",
    "    elif np.any(np.array(row.values[:-2]) == 'WARNING'):\n",
    "        sample_status.loc[index,'status'] = 'WARNING'\n",
    "        \n",
    "failed_reals = sample_status.loc[data['group']=='Real'].loc[sample_status['status']=='FAIL']['samples']\n",
    "\n",
    "print(f'\\tFAILED {len(failed_reals)} real samples:\\n')\n",
    "for sample in failed_reals:\n",
    "    print('\\t' + sample)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "group_colors = dict(Real='#80b1d3', Empty='#ffffb3', H2O='#bebada', Negative='#fb8072', Positive='#8dd3c7')\n",
    "\n",
    "layout = go.Layout(\n",
    "    title = 'Sample group sizes',\n",
    "    xaxis = go.layout.XAxis(\n",
    "        title = 'Group',\n",
    "        showticklabels=True),\n",
    "    yaxis = go.layout.YAxis(\n",
    "        title = 'Number of samples'\n",
    "    ),\n",
    "    margin = go.layout.Margin(t = 50),\n",
    "    template=\"plotly_white\"\n",
    "    \n",
    ")\n",
    "\n",
    "colors = ['#3182bd'] + list(group_colors.values())\n",
    "fig = go.Figure()\n",
    "\n",
    "# Add traces\n",
    "fig.add_trace(go.Bar(x=groupsize['group'], y=groupsize['size'],\n",
    "                    marker=dict(color=colors)))\n",
    "\n",
    "fig.update_layout(layout)\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "colorscale=[[0, \"red\"], [0.111111111111, \"yellow\"], [0.222222222222, \"green\"], \n",
    "          [0.333333333333, \"rgb(141,211,199)\"], [0.444444444444, \"rgb(255,255,179)\"], \n",
    "          [0.555555555556, \"rgb(190,186,218)\"], [0.666666666667, \"rgb(190,186,218)\"], \n",
    "          [0.777777777778, \"rgb(128,177,211)\"], [0.888888888889, \"rgb(128,177,211)\"], \n",
    "          [1, \"rgb(49,54,149)\"]]\n",
    "passcolordicts = {'FAIL': 0., 'WARNING': 1./9., 'PASS': 2./9.}\n",
    "groupcolordicts = {'Real': 7./9., 'Empty': 6./9., 'H2O': 5./9., 'Negative': 4./9., 'Positive': 3./9.}\n",
    "colors = np.array([[passcolordicts[v] for v in flags[s].values] for s in list(flags.columns[1:])])\n",
    "colors = np.concatenate([np.array([[groupcolordicts[group] for group in data['group']]]), colors])\n",
    "colors = np.concatenate([np.array([[8./9. for s in list(flags.columns)]]), colors.T])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "flags2 = flags.rename(columns={'coverage': 'Coverage', 'r1_basequal': 'Base quality R1', 'r2_basequal': 'Base quality R2',\n",
    "                              'rejreads': 'Rejected reads', 'alnreads': 'Aligned reads', 'insertsize': 'Insert size', \n",
    "                               'consensus_N': 'Consensus<br>undetermined',\n",
    "                               'consensus_lcbases': 'Consensus<br>ambiguousity', 'consensus_diffbases': 'Consensus<br>sequence',\n",
    "                              'snvs': 'Number of<br>SNVs', 'snvs_majority': 'Majority<br>SNVs'})\n",
    "from copy import deepcopy\n",
    "flags3 = deepcopy(flags2)\n",
    "flags3.loc[np.where(np.array(['pos' in  f for f in flags3['Sample']]))[0],'Sample'] = ['_'.join(a.split('_', 3)[:2]) for a in flags3.loc[np.where(np.array(['pos' in  f for f in flags3['Sample']]))[0],'Sample'].tolist()]\n",
    "flags3.loc[np.where(np.array(['EMPTY' in  f for f in flags3['Sample']]))[0],'Sample'] = ['_'.join(a.split('_', 3)[:2]) for a in flags3.loc[np.where(np.array(['EMPTY' in  f for f in flags3['Sample']]))[0],'Sample'].tolist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "descriptions = {'Sample':\"\",\n",
    "                'Coverage': f'Third quartile of the coverage across the whole genome.<br>WARNING if < {coverage_3q_min_warning}<br>FAIL if < {coverage_3q_min_fail}',\n",
    "               'Base quality R1': 'FastQC per base quality flag for read 1',\n",
    "               'Base quality R2': 'FastQC per base quality flag for read 2',\n",
    "               'Rejected reads': f'Percentage of reads rejected by Prinseq.<br>WARNING if > {rejupper_bound_warning}%<br>FAIL if > {rejupper_bound_fail}%',\n",
    "               'Aligned reads': f'Percentage of kept reads that were aligned.<br>FAIL if < {alnlower_bound}%',\n",
    "               'Insert size': f'Minimum and maximum estimated insert sizes.<br>FAIL if minimum < {inslower_bound} or maximum > {insupper_bound}',\n",
    "                'Consensus<br>undetermined': f'Number of bases considered undetermined, i.e., with support in less than 5 reads.<br>FAIL if > {undupper_bound}',\n",
    "               'Consensus<br>ambiguosity': f'Number of bases considered ambiguous, i.e., with support in less than 50 reads.<br>FAIL if > {ambupper_bound}',\n",
    "               'Consensus<br>sequence': f'Number of bases different from the reference sequence. \"N\" calls not included.<br>WARNING if > {diffupper_bound_warning}<br>FAIL if > {diffupper_bound_fail}',\n",
    "               'Number of<br>SNVs': f'Number of single-nucleotide variants<br>FAIL if > {snvsupper_bound}',\n",
    "               'Majority<br>SNVs': f'Number of consensus single-nucleotide variants<br>FAIL if 10x larger than the consensus sequence'}\n",
    "\n",
    "hover=[]\n",
    "hover.append(['<b>' + desc + '</b>' for desc in list(descriptions.values())])\n",
    "for i in range(flags_values.shape[0]):\n",
    "    row = [flags_values.iloc[i]['Sample']]\n",
    "    for j in range(1, flags_values.shape[1]):\n",
    "        val = str(flags_values.iloc[i, j])\n",
    "        if flags_values.columns[j] == 'rejreads' or flags_values.columns[j] == 'alnreads':\n",
    "            val = f\"{flags_values.iloc[i, j]:.2f}\"\n",
    "        row.append('<b>' + list(descriptions.values())[j] + '</b>' + '<br><br>' +\n",
    "                   '<b>' + str(flags_values.iloc[i, 0]) + ': </b>' + val + '<br><br>' + \n",
    "                  'Coverage 3Q: ' + str(flags_values.iloc[i]['coverage']) + '<br>' +\n",
    "                  f\"% Rejected reads: {flags_values.iloc[i]['rejreads']:.2f} <br>\" +\n",
    "                  f\"% Aligned reads: {flags_values.iloc[i]['alnreads']:.2f}\")\n",
    "                      \n",
    "    hover.append(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "flags_values.to_csv(output_qafile, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "table_simple = ff.create_table(flags3, text=hover, hoverinfo='text', )\n",
    "#                             hovertemplate=\"<b>%{text}</b><br><br>%{y}<br><br>\" +\n",
    "#                         \"<extra></extra>\")\n",
    "table_simple.update_layout(autosize=False,\n",
    "    width=1400,)\n",
    "d = table_simple.to_dict()\n",
    "d['data'][0]['z'] = colors\n",
    "d['data'][0]['colorscale'] = colorscale\n",
    "\n",
    "pio.show(d, config={'displayModeBar': False})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "delta_qs = coverage_quartiles.T.iloc[1:].T.diff(axis=1)\n",
    "delta_qs['sample'] = coverage_quartiles['sample']\n",
    "delta_qs['25%'] = coverage_quartiles['25%']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "layout = go.Layout(\n",
    "    title = 'Coverage quartiles per sample',\n",
    "    xaxis = go.layout.XAxis(\n",
    "        title = 'Samples',\n",
    "        showticklabels=False),\n",
    "    yaxis = go.layout.YAxis(\n",
    "        title = 'Coverage'\n",
    "    ),\n",
    "    template=\"plotly_white\",\n",
    "    barmode='relative',\n",
    "    )\n",
    "\n",
    "colors = ['#6baed6', '#3182bd', '#08519c']\n",
    "fig = go.Figure()\n",
    "\n",
    "# Add traces\n",
    "fig.add_trace(go.Bar(x=coverage_quartiles['sample'], y=delta_qs['25%'],\n",
    "                    #mode='lines',\n",
    "                     text=coverage_quartiles['25%'],\n",
    "                     hovertemplate=\"(%{x}, %{text})<extra></extra>\",\n",
    "                    name='1Q', marker=dict(color=colors[0])))\n",
    "fig.add_trace(go.Bar(x=coverage_quartiles['sample'], y=delta_qs['50%'],\n",
    "                    #mode='lines',\n",
    "                      text=coverage_quartiles['50%'],\n",
    "                     hovertemplate=\"(%{x}, %{text})<extra></extra>\",\n",
    "                    name='Median', marker=dict(color=colors[1])))\n",
    "fig.add_trace(go.Bar(x=coverage_quartiles['sample'], y=delta_qs['75%'],\n",
    "                    #mode='lines',\n",
    "                      text=coverage_quartiles['75%'],\n",
    "                     hovertemplate=\"(%{x}, %{text})<extra></extra>\",\n",
    "                    name='3Q', marker=dict(color=colors[2])))\n",
    "fig.add_trace(go.Scatter(x=coverage_quartiles['sample'], y=np.ones((coverage_quartiles.shape[0])) * coverage_3q_min_warning,\n",
    "                    mode='lines',\n",
    "                    name='Soft lower bound\\nfor 3Q', marker=dict(color='#CCCC00'), line = dict(dash='dot')))\n",
    "fig.add_trace(go.Scatter(x=coverage_quartiles['sample'], y=np.ones((coverage_quartiles.shape[0])) * coverage_3q_min_fail,\n",
    "                    mode='lines',\n",
    "                    name='Hard lower bound\\nfor 3Q', marker=dict(color='red'), line = dict(dash='dot')))\n",
    "\n",
    "fig.update_layout(layout)\n",
    "\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "group_percentiles = dict(\n",
    "    Real=pd.DataFrame({'25%':coverage.loc[data['group']=='Real'].quantile(q=0.25, axis=0),\n",
    "                '50%':coverage.loc[data['group']=='Real'].quantile(q=0.5, axis=0),\n",
    "                '75%':coverage.loc[data['group']=='Real'].quantile(q=0.75, axis=0)}),\n",
    "    Empty=pd.DataFrame({'25%':coverage.loc[data['group']=='Empty'].quantile(q=0.25, axis=0),\n",
    "                '50%':coverage.loc[data['group']=='Empty'].quantile(q=0.5, axis=0),\n",
    "                '75%':coverage.loc[data['group']=='Empty'].quantile(q=0.75, axis=0)}),\n",
    "    H2O=pd.DataFrame({'25%':coverage.loc[data['group']=='H2O'].quantile(q=0.25, axis=0),\n",
    "                '50%':coverage.loc[data['group']=='H2O'].quantile(q=0.5, axis=0),\n",
    "                '75%':coverage.loc[data['group']=='H2O'].quantile(q=0.75, axis=0)}),\n",
    "    Negative=pd.DataFrame({'25%':coverage.loc[data['group']=='Negative'].quantile(q=0.25, axis=0),\n",
    "                '50%':coverage.loc[data['group']=='Negative'].quantile(q=0.5, axis=0),\n",
    "                '75%':coverage.loc[data['group']=='Negative'].quantile(q=0.75, axis=0)}),\n",
    "    Positive=pd.DataFrame({'25%':coverage.loc[data['group']=='Positive'].quantile(q=0.25, axis=0),\n",
    "                '50%':coverage.loc[data['group']=='Positive'].quantile(q=0.5, axis=0),\n",
    "                '75%':coverage.loc[data['group']=='Positive'].quantile(q=0.75, axis=0)}))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "colors = ['#6baed6', '#3182bd', '#08519c']\n",
    "\n",
    "groups = data['group'].unique()\n",
    "vals = []\n",
    "for i, g in enumerate(groups):\n",
    "    a = [False, False, False] * len(groups)\n",
    "    a[i*3:i*3+3] = [True, True, True]\n",
    "    vals.append(a)\n",
    "\n",
    "# Add traces\n",
    "fig = go.Figure()\n",
    "for i, group in enumerate(groups):\n",
    "    for j, p in enumerate(['25%', '50%', '75%']):\n",
    "        fig.add_trace(go.Scatter(x=group_percentiles[group].index, y=group_percentiles[group][p],\n",
    "                            mode='lines',\n",
    "                            name=['1st quart.', 'Median', '3rd quart.'][j], marker=dict(color=colors[j]),\n",
    "                            visible=False if i > 0 else True))\n",
    "\n",
    "updatemenus = list([\n",
    "    dict(active=0,\n",
    "         showactive = True,\n",
    "         buttons=list([dict(label=group,\n",
    "                           method=\"update\",\n",
    "                           args =[{\"visible\": vals[i]}])\n",
    "            for i, group in enumerate(groups)\n",
    "            ]),\n",
    "            direction=\"down\",\n",
    "            pad={\"r\": 10, \"t\": 10},\n",
    "            x=0.05,\n",
    "            xanchor=\"left\",\n",
    "            y=1.12,\n",
    "            yanchor=\"top\")])\n",
    "\n",
    "layout = go.Layout(\n",
    "    title = 'Coverage quartiles per position per group',\n",
    "    xaxis = go.layout.XAxis(\n",
    "        title = 'Position',\n",
    "        showticklabels=True),\n",
    "    yaxis = go.layout.YAxis(\n",
    "        title = 'Coverage'\n",
    "    ),\n",
    "    template=\"plotly_white\",\n",
    "    updatemenus=updatemenus\n",
    ")\n",
    "\n",
    "fig.update_layout(layout)\n",
    "# Add annotation\n",
    "fig.update_layout(\n",
    "    annotations=[\n",
    "        dict(text=\"Group:\", showarrow=False,\n",
    "        x=0., y=1.085, yref=\"paper\", align=\"left\")\n",
    "    ]\n",
    ")\n",
    "\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prinseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "layout = go.Layout(\n",
    "    title = 'Reads rejected by Prinseq',\n",
    "    xaxis = go.layout.XAxis(\n",
    "        title = 'Samples',\n",
    "        showticklabels=False),\n",
    "    yaxis = go.layout.YAxis(\n",
    "        title = 'Reads rejected (%)'\n",
    "    ),\n",
    "    template=\"plotly_white\"\n",
    "    \n",
    ")\n",
    "\n",
    "# fig = go.Figure(data=[{'y': coverage_quartiles['coverage'], 'x': coverage_quartiles['sample']}], layout=layout)\n",
    "# iplot(fig)\n",
    "colors = ['#3182bd']\n",
    "fig = go.Figure()\n",
    "\n",
    "# Add traces\n",
    "fig.add_trace(go.Bar(x=rejreads['sample'], y=rejreads['All'],  name=\"Observed\",\n",
    "                    marker=dict(color=colors[0]),\n",
    "                    text = rejreads[\"Absolute\"].values,\n",
    "                    hoverinfo = 'y+text',\n",
    "                    hovertemplate=\n",
    "                        \"<b>%{x}</b><br><br>\" +\n",
    "                        \"Percentage: %{y:.2f}%<br>\" +\n",
    "                        \"Absolute: %{text}<br>\"\n",
    "                        \"<extra></extra>\",\n",
    "                    ))\n",
    "fig.add_trace(go.Scatter(x=rejreads['sample'], y=np.ones((rejreads.shape[0])) * rejupper_bound_warning,\n",
    "                    mode='lines',\n",
    "                    name='Soft upper bound', marker=dict(color='#CCCC00'), line = dict(dash='dot')))\n",
    "fig.add_trace(go.Scatter(x=rejreads['sample'], y=np.ones((rejreads.shape[0])) * rejupper_bound_fail,\n",
    "                    mode='lines',\n",
    "                    name='Hard upper bound', marker=dict(color='red'), line = dict(dash='dot')))\n",
    "fig.update_layout(layout)\n",
    "fig.show()\n",
    "# fig = px.line(coverage_quartiles, x=\"sample\", y=\"coverage\",\n",
    "#               color='percentile', hover_name=\"sample\", template=\"plotly_white\")\n",
    "\n",
    "# iplot(fig)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "layout = go.Layout(\n",
    "    title = 'Reads aligned by BWA',\n",
    "    xaxis = go.layout.XAxis(\n",
    "        title = 'Samples',\n",
    "        showticklabels=False),\n",
    "    yaxis = go.layout.YAxis(\n",
    "        title = 'Reads aligned (%)'\n",
    "    ),\n",
    "    template=\"plotly_white\"\n",
    "    \n",
    ")\n",
    "\n",
    "colors = ['#3182bd']\n",
    "fig = go.Figure()\n",
    "\n",
    "# Add traces\n",
    "fig.add_trace(go.Bar(x=alnreads['sample'], y=alnreads['All'].values, name=\"Observed\",\n",
    "                    marker=dict(color=colors[0]),\n",
    "                    text = alnreads[\"Absolute\"].values,\n",
    "                    hoverinfo = 'y+text',\n",
    "                    hovertemplate=\n",
    "                        \"<b>%{x}</b><br><br>\" +\n",
    "                        \"Percentage: %{y:.2f}%<br>\" +\n",
    "                        \"Absolute: %{text}<br>\"\n",
    "                        \"<extra></extra>\",\n",
    "                    ))\n",
    "fig.add_trace(go.Scatter(x=alnreads['sample'], y=np.ones((alnreads.shape[0])) * 90,\n",
    "                    mode='lines',\n",
    "                    name='Lower bound', marker=dict(color='red'), line = dict(dash='dot')))\n",
    "\n",
    "fig.update_layout(layout)\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "delta_is = bwa_insert_dist.T.iloc[1:].T.diff(axis=1)\n",
    "delta_is['sample'] = bwa_insert_dist['sample']\n",
    "delta_is['min'] = bwa_insert_dist['min']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "layout = go.Layout(\n",
    "    title = 'Insert sizes estimated by BWA',\n",
    "    xaxis = go.layout.XAxis(\n",
    "        title = 'Samples',\n",
    "        showticklabels=False),\n",
    "    yaxis = go.layout.YAxis(\n",
    "        title = 'Insert sizes',\n",
    "        range = [0, 600]\n",
    "    ),\n",
    "    template=\"plotly_white\",\n",
    "    barmode='stack'\n",
    ")\n",
    "\n",
    "colors = ['#6baed6', '#3182bd', '#08519c']\n",
    "fig = go.Figure()\n",
    "\n",
    "# Add traces\n",
    "fig.add_trace(go.Bar(x=bwa_insert_dist['sample'], y=delta_is['min'],\n",
    "                    #mode='lines',\n",
    "                     text=bwa_insert_dist['min'],\n",
    "                     hovertemplate=\"(%{x}, %{text})<extra></extra>\",\n",
    "                    name='Min', marker=dict(color=colors[0])))\n",
    "fig.add_trace(go.Bar(x=bwa_insert_dist['sample'], y=delta_is['mean'],\n",
    "                    #mode='lines',\n",
    "                     text=bwa_insert_dist['mean'],\n",
    "                     hovertemplate=\"(%{x}, %{text})<extra></extra>\",\n",
    "                    name='Mean', marker=dict(color=colors[1])))\n",
    "fig.add_trace(go.Bar(x=bwa_insert_dist['sample'], y=delta_is['max'],\n",
    "                    #mode='lines',\n",
    "                     text=bwa_insert_dist['max'],\n",
    "                     hovertemplate=\"(%{x}, %{text})<extra></extra>\",\n",
    "                    name='Max', marker=dict(color=colors[2])))\n",
    "\n",
    "fig.add_trace(go.Scatter(x=rejreads['sample'], y=np.ones((rejreads.shape[0])) * 300,\n",
    "                    mode='lines',\n",
    "                    name='Lower bound', marker=dict(color='red'), line = dict(dash='dot')))\n",
    "fig.add_trace(go.Scatter(x=rejreads['sample'], y=np.ones((rejreads.shape[0])) * 500,\n",
    "                    mode='lines',\n",
    "                    name='Upper bound', marker=dict(color='red'), line = dict(dash='dot')))\n",
    "\n",
    "fig.update_layout(layout)\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Consensus sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "layout = go.Layout(\n",
    "    title = 'Number of undetermined (N) bases in consensus',\n",
    "    xaxis = go.layout.XAxis(\n",
    "        title = 'Samples',\n",
    "        showticklabels=False),\n",
    "    yaxis = go.layout.YAxis(\n",
    "        title = 'Number of N calls'\n",
    "    ),\n",
    "    template=\"plotly_white\"\n",
    "    \n",
    ")\n",
    "\n",
    "colors = ['#3182bd']\n",
    "fig = go.Figure()\n",
    "\n",
    "# Add traces\n",
    "fig.add_trace(go.Bar(x=nbases['sample'], y=nbases['All'], name=\"Observed\",\n",
    "                    marker=dict(color=colors[0])))\n",
    "fig.add_trace(go.Scatter(x=nbases['sample'], y=np.ones((nbases.shape[0])) * undupper_bound,\n",
    "                    mode='lines',\n",
    "                    name='Upper bound', marker=dict(color='red'), line = dict(dash='dot')))\n",
    "\n",
    "fig.update_layout(layout)\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "layout = go.Layout(\n",
    "    title = 'Number of ambiguous (lowercase) bases in consensus',\n",
    "    xaxis = go.layout.XAxis(\n",
    "        title = 'Samples',\n",
    "        showticklabels=False),\n",
    "    yaxis = go.layout.YAxis(\n",
    "        title = 'Number of lowercase calls'\n",
    "    ),\n",
    "    template=\"plotly_white\"\n",
    "    \n",
    ")\n",
    "\n",
    "colors = ['#3182bd']\n",
    "fig = go.Figure()\n",
    "\n",
    "# Add traces\n",
    "fig.add_trace(go.Bar(x=lcbases['sample'], y=lcbases['All'], name=\"Observed\",\n",
    "                    marker=dict(color=colors[0])))\n",
    "fig.add_trace(go.Scatter(x=lcbases['sample'], y=np.ones((lcbases.shape[0])) * ambupper_bound,\n",
    "                    mode='lines',\n",
    "                    name='Upper bound', marker=dict(color='red'), line = dict(dash='dot')))\n",
    "\n",
    "fig.update_layout(layout)\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "layout = go.Layout(\n",
    "    title = 'Number of different bases* between consensus and reference<br><span style=\"font-size:12\">*not including \"N\"</span>',\n",
    "    xaxis = go.layout.XAxis(\n",
    "        title = 'Samples',\n",
    "        showticklabels=False),\n",
    "    yaxis = go.layout.YAxis(\n",
    "        title = 'Number of bases',\n",
    "        range = [0, diffupper_bound_fail + 10]\n",
    "    ),\n",
    "    template=\"plotly_white\"\n",
    "    \n",
    ")\n",
    "colors = ['#3182bd']\n",
    "fig = go.Figure()\n",
    "\n",
    "# Add traces\n",
    "fig.add_trace(go.Bar(x=diffbases['sample'], y=diffbases['All'], name=\"Observed\",\n",
    "                     marker=dict(color=colors[0])))\n",
    "fig.add_trace(go.Scatter(x=diffbases['sample'], y=np.ones((diffbases.shape[0])) * diffupper_bound_warning,\n",
    "                    mode='lines',\n",
    "                    name='Soft upper bound', marker=dict(color='#CCCC00'), line = dict(dash='dot')))\n",
    "\n",
    "fig.add_trace(go.Scatter(x=diffbases['sample'], y=np.ones((diffbases.shape[0])) * diffupper_bound_fail,\n",
    "                    mode='lines',\n",
    "                    name='Hard upper bound', marker=dict(color='red'), line = dict(dash='dot')))\n",
    "fig.update_layout(layout)\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ShoRAH"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "layout = go.Layout(\n",
    "    title = 'SNVs called by ShoRAH (posterior probability > 80%)',\n",
    "    xaxis = go.layout.XAxis(\n",
    "        title = 'Samples',\n",
    "        showticklabels=False),\n",
    "    yaxis = go.layout.YAxis(\n",
    "        title = 'Number of SNVs'\n",
    "    ),\n",
    "    template=\"plotly_white\"\n",
    "    \n",
    ")\n",
    "\n",
    "colors = ['#3182bd']\n",
    "fig = go.Figure()\n",
    "\n",
    "# Add traces\n",
    "fig.add_trace(go.Bar(x=filtered_snvs['sample'], y=filtered_snvs['All'], name=\"Observed\",\n",
    "                     marker=dict(color=colors[0])))\n",
    "fig.add_trace(go.Scatter(x=filtered_snvs['sample'], y=np.ones((filtered_snvs.shape[0])) * 2000,\n",
    "                    mode='lines',\n",
    "                    name='Upper bound', marker=dict(color='red'), line = dict(dash='dot')))\n",
    "\n",
    "fig.update_layout(layout)\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "layout = go.Layout(\n",
    "    title = 'Number of SNVs from filtered ShoRAH-based majority vote',\n",
    "    xaxis = go.layout.XAxis(\n",
    "        title = 'Samples',\n",
    "        showticklabels=False),\n",
    "    yaxis = go.layout.YAxis(\n",
    "        title = 'Number of SNVs'\n",
    "    ),\n",
    "    template=\"plotly_white\",\n",
    "    yaxis_type=\"log\",\n",
    "    \n",
    ")\n",
    "\n",
    "colors = ['#3182bd']\n",
    "fig = go.Figure()\n",
    "\n",
    "# Add traces\n",
    "fig.add_trace(go.Bar(x=maj_snvs['sample'], y=maj_snvs['All'], name=\"Observed\",\n",
    "                    marker=dict(color=colors[0])))\n",
    "fig.add_trace(go.Scatter(x=maj_snvs['sample'], y=majupper_bounds,\n",
    "                    mode='markers',\n",
    "                    name='Upper bound', marker=dict(color='red'), line = dict(dash='dot')))\n",
    "\n",
    "fig.update_layout(layout)\n",
    "fig.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "conda-qa_report",
   "language": "python",
   "name": "conda-qa_report"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
